### DISCUSSION TITLE
Cut Development Release 0.24

### DISCUSSION THREAD

**Discussion by diesalbla** [http4s/http4s#7191] [General]
Currently, development of Http4S is done in two branches, `series/0.23` and `main`. The former is being released in the `0.23.x` releases, the last one `0.23.22` from today, and the first one `0.23.0` is from almost _two years ago_. In the meantime, the `main` branch intended to be released as `1.0` has been in development for three years since [1.0.0-M1](https://github.com/http4s/http4s/releases/tag/v1.0.0-M1). The current git-diff between the two branches gives a total of 4315 insertions, and 6181 deletions. All of these line changes include binary-incompatible changes in `main` that are not being widely used. 

Unfortunately, although the `main` branch has been in development, the design of that branch is not completed and there are issues to be addressed. To narrow some of this gap, and to make the improvements in `main` available more generally before the `1.0` version is released, there should be a `0.24.x` series that includes some binary incompatible changes. In particular, it should include the improvements to the `Entity` representation, which improved microbenchmark measurements. It could also include some of the code deletions that are included in `main`.

**Comment by bplommer** [comment]
I feel the same way but am reluctant to try and impose the extra maintenance burden on @rossabaker

**Reply by diesalbla** [reply]
> I feel the same way but am reluctant to try and impose the extra maintenance burden on @rossabaker

- The additional maintenance burden of a new `0.24` series could be compensated if the `0.23` branch were to be closed for code changes and marked as "fixes-only".
- Although a `0.24` may be a bit more maintenance burden, by spreading some of the changes in the `main` branch sooner, it would also help to detect any lurking problems earlier.
- Is Ross Baker the only maintainer of the Http4S library? Who else takes care of maintenance? Which tasks could be delegated?

**Comment by hamnis** [comment]
One reason for going to 0.24 could be that its starts to get increasingly painful to merges. I think @armanbilge and @danicheg can certainly attest to that. Having two increasingly divergent branches is a maintenance burden.

Although that is a problem, we have been constantly dropping milestone releases, which users may try. Is main in a releasable state as a 0.24 series now, or do we need to stabilize something?

Could we call main as 1.0 and rather focus on a 2.0 with cats-uri and other breaking changes? Why is 1.0 so scary?

**Comment by danicheg** [comment]
> Currently, development of Http4S is done in two branches, series/0.23 and main. The former is being released in the 0.23.x releases, the last one 0.23.22 from today, and the first one 0.23.0 is from almost two years ago. In the meantime, the main branch intended to be released as 1.0 has been in development for three years since [1.0.0-M1](https://github.com/http4s/http4s/releases/tag/v1.0.0-M1).

It's a weak justification, but... `Netty` has been doing this for 7 years. `4.1.0.Final` (May 25, 2016) -> `4.1.94.Final` (Jun 20, 2023). They even have a fancy postfix `Final`!

Joking aside, I am also a little worried that migration `0.23` -> `1.0` could be very tricky already. But my two cents is that bringing `0.24` won't be a silver bullet. Perhaps we should revisit our `1.0` milestone and work on stabilizing it to make the release happen sooner.

**Comment by rossabaker** [comment]
It's no longer fair to name me specifically for maintenance burden, particularly when I've been a peripheral figure in the last few releases.  I'd rather address it as an engineer at a company with seven Scala teams.  We're in the midst of an engineering garbage collection pause, where we finally get off cats-effect-2 and friends, including http4s-0.21.  Breaking changes are terrible things.  Beyond that, there's a significant downstream ecosystem of integrations built on top of 0.23.  Every time we introduce a breaking release series as production, that entire ecosystem has to support two, and/or every adopter like my employer hits another GC pause.  This is a heavy decision.

I don't see any practical difference between labeling a release 0.24.0 and a 1.0.0.  It's going to break all the things, and the time we break all the things is:

1. when it's compelling enough to justify the pain
2. when it's been batched up enough we won't have to do it again soon

I don't think the first condition is close to true yet.  There's still the transformers so many people hate.  There's still the URI, which is the thing people should hate as much as they hate the transformers.  There is still the import tax that daunts new users.[^1]    The new entity model is smarter, but unless you want to argue that Techempower isn't horseshit, is it sufficiently better to disrupt every library and user downstream?

Now, 1.0 is proceeding at a crawl because we've done such a good job improving 0.23 while stable.  To reach those conditions to make 1.0 worthwhile, we could back off enhancements on 0.23 and make it bugfixes only (roughly @danicheg's point above).  The branches will diverge further, so each merge may get harder, but fewer merges could lower the overall burden.  But we would want:

- A clearer roadmap of what 1.0 means.  Breaking releases only happen around here when frustration builds to the point that someone reluctantly wears the project manager hat.
  - Needs an issue scrub
  - Needs clearer milestone tags
  - Needs some discipline rather than ad hoc scratching of itches
- A clearer definition of what continues to happen on 0.23 and what doesn't.
- Some nonbinding show of hands that people are prepared to push.  Whatever improvements are on an unreleased main don't matter, so time becomes of the essence in a way it is not today.

[^1]: I don't yet know how to make Cats URI happen without making the import tax worse

**Reply by Daenyth** [reply]
> I don't yet know how to make Cats URI happen without making the import tax worse

In what way?
And could it be helped with an alias in the http4s package?

**Reply by armanbilge** [reply]
See related discussion in:
- https://github.com/http4s/http4s/discussions/6696
